service/matching/tasklist/task_completer.go

```diff
@@ -0,0 +1,204 @@
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/uber/cadence/client/history"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/backoff"
+	"github.com/uber/cadence/common/cache"
+	"github.com/uber/cadence/common/cluster"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/persistence"
+	"github.com/uber/cadence/common/types"
+)
+
+type (
+	taskCompleterImpl struct {
+		domainCache     cache.DomainCache
+		taskListID      *Identifier
+		clusterMetadata cluster.Metadata
+		historyService  history.Client
+		scope           metrics.Scope
+		logger          log.Logger
+		throttleRetry   *backoff.ThrottleRetry
+	}
+)
+
+// DomainIsActiveInThisClusterError type
+type DomainIsActiveInThisClusterError struct {
+	Message string
+}
+
+// Implement the Error method for DomainIsActiveInThisClusterError
+func (e *DomainIsActiveInThisClusterError) Error() string {
+	return e.Message
+}
+
+var (
+	errWorkflowExecutionInfoIsNil      = errors.New("workflow execution info is nil")
+	errTaskTypeNotSupported            = errors.New("task type not supported")
+	errTaskNotStarted                  = errors.New("task not started")
+	errDomainIsActive                  = &DomainIsActiveInThisClusterError{Message: "domain is active"}
+	historyServiceOperationRetryPolicy = common.CreateTaskCompleterRetryPolicy()
+)
+
+func newTaskCompleter(tlMgr *taskListManagerImpl, retryPolicy backoff.RetryPolicy) TaskCompleter {
+	return &taskCompleterImpl{
+		domainCache:     tlMgr.domainCache,
+		historyService:  tlMgr.historyService,
+		taskListID:      tlMgr.taskListID,
+		clusterMetadata: tlMgr.clusterMetadata,
+		scope:           tlMgr.scope,
+		logger:          tlMgr.logger,
+		throttleRetry: backoff.NewThrottleRetry(
+			backoff.WithRetryPolicy(retryPolicy),
+			backoff.WithRetryableError(isRetryableError),
+		),
+	}
+}
+
+func (tc *taskCompleterImpl) CompleteTaskIfStarted(ctx context.Context, task *InternalTask) error {
+	op := func() (err error) {
+		domainEntry, err := tc.domainCache.GetDomainByID(task.Event.TaskInfo.DomainID)
+		if err != nil {
+			return fmt.Errorf("unable to fetch domain from cache: %w", err)
+		}
+
+		if _, err = domainEntry.IsActiveIn(tc.clusterMetadata.GetCurrentClusterName()); err == nil {
+			return errDomainIsActive
+		}
+
+		req := &types.HistoryDescribeWorkflowExecutionRequest{
+			DomainUUID: task.Event.TaskInfo.DomainID,
+			Request: &types.DescribeWorkflowExecutionRequest{
+				Domain: task.domainName,
+				Execution: &types.WorkflowExecution{
+					WorkflowID: task.Event.WorkflowID,
+					RunID:      task.Event.RunID,
+				},
+			},
+		}
+
+		workflowExecutionResponse, err := tc.historyService.DescribeWorkflowExecution(ctx, req)
+
+		if errors.As(err, new(*types.EntityNotExistsError)) {
+			return nil
+		} else if err != nil {
+			return fmt.Errorf("unable to fetch workflow execution from the history service: %w", err)
```

Comment:
Mind adding a log here with some info such as the workflow ID and so fourth, this would near certainly be a bug, but in cases were we see unstuck tasklists it's forseeable. We probably wnat to know about such events.

Reply:
The case here is that it could take a while for the task list manager in the standby side to actually attempt to process the task (although it's unlikely), and based on the retention period for the workflow, we could be trying to get information about a workflow that is not present in the db anymore. Does that make sense?

Reply:
I would put it as a warn at least, I'd be pretty surprised for the average retention (7 days) to exceed the duration of somehting in the tasks table